packages/axelar-wasm-std/src/permission_control.rs

```diff
@@ -0,0 +1,283 @@
+        Admin = 2,
+        Governance = 4,
+        Elevated = (Permission::Admin | Permission::Governance).bits(),
+        Any = (Permission::NoPrivilege | Permission::Elevated).bits(),
+    }
+}
+
+#[derive(thiserror::Error, Debug, PartialEq)]
+pub enum Error {
+    #[error("sender is not allowed to perform this action")]
+    PermissionDenied,
+}
+
+/// Ensure that the sender of a message has the correct permissions to perform following actions.
+/// Returns an error if not.
+/// # Example
+/// ```
+/// # use cosmwasm_std::testing::mock_dependencies;
+/// use cosmwasm_std::Addr;
+/// use axelar_wasm_std::ensure_permission;
+/// use axelar_wasm_std::permission_control::Permission;
+///# use axelar_wasm_std::permission_control::Error;
+///
+///# fn main() -> Result<(),Box<dyn std::error::Error>>{
+///# use axelar_wasm_std::permission_control;
+///# let mut deps = mock_dependencies();
+///# let deps = deps.as_mut();
+/// let admin = Addr::unchecked("admin");
+/// let governance = Addr::unchecked("governance");
+///
+/// // set these before checking permissions
+/// permission_control::set_admin(deps.storage, &admin)?;
+/// permission_control::set_governance(deps.storage, &governance)?;
+///
+/// ensure_permission!(Permission::Elevated, deps.storage, &admin);
+/// ensure_permission!(Permission::Elevated, deps.storage, &governance);
+///
+/// do_something();
+/// # Ok(())}
+///
+/// #    fn do_something() {}
+/// ```
+#[macro_export]
+macro_rules! ensure_permission {
+    ($permission_variant:expr, $storage:expr, $sender:expr) => {
+        let permission = $crate::flagset::FlagSet::from($permission_variant);
+        if !permission.contains($crate::permission_control::Permission::Any)
+            && (*permission
+                & *(error_stack::ResultExt::change_context(
+                    $crate::permission_control::sender_role($storage, $sender),
+                    $crate::permission_control::Error::PermissionDenied,
+                ))?)
+            .is_empty()
+        {
+            return Err($crate::permission_control::Error::PermissionDenied.into());
+        }
+    };
+}
+
+/// This macro should be used as a marker to signify that the call is deliberately without checks
+///
+/// # Example
+/// ```
+///# fn main() -> Result<(),Box<dyn std::error::Error>>{
+///# use axelar_wasm_std::{ensure_any_permission};
+/// ensure_any_permission!();
+/// do_something();
+/// # Ok(())}
+///
+/// #    fn do_something() {}
+/// ```
+#[macro_export]
+macro_rules! ensure_any_permission {
+    () => {};
+}
+
+const ADMIN: Item<Addr> = Item::new("permission_control_contract_admin_addr");
+
+const GOVERNANCE: Item<Addr> = Item::new("permission_control_governance_addr");
+
+pub fn set_admin(storage: &mut dyn cosmwasm_std::Storage, addr: &Addr) -> StdResult<()> {
+    ADMIN.save(storage, addr)
+}
+
+pub fn set_governance(storage: &mut dyn cosmwasm_std::Storage, addr: &Addr) -> StdResult<()> {
+    GOVERNANCE.save(storage, addr)
+}
```

Comment:
How does this work with multiple contracts? Do they each have their own storage, and the db somehow prefixes by contract address or some such?

Reply:
yes, the storage is namespaced to the contract